openghg/datapack/_obspack.py

```diff
@@ -0,0 +1,1385 @@
+    See define_obs_types() for different obs types currently supported.
+
+    Example:
+    >    site,inlet,species,obs_type
+    >    tac,185m,co2,surface-insitu
+    >    bsd,42m,ch4,surface-insitu
+    >    bsd,100m-250m,ch4,surface-insitu
+
+    Args:
+        filename: Filename containing search parameters and obs_type column.
+    Returns:
+        pandas.DataFrame: Input file opened as a csv file using pd.read_csv(...)
+
+    TODO: Add checks for expected columns in input search file as required
+    """
+    search_df = pd.read_csv(filename)
+    return search_df
+
+
+def retrieve_data(
+    filename: pathType | None = None, search_df: pd.DataFrame | None = None, store: str | None = None
+) -> list:
+    """
+    Use search parameters to get data from object store. This expects either a filename for an input
+    file containing search parameters (see read_input_file() for more details) or a DataFrame containing
+    the search parameters.
+
+    Args:
+        filename: Filename containing search parameters.
+        search_df: pandas DataFrame containing search parameters
+        store: Name of specific object store to use to search for data
+    Returns:
+        list [StoredData]: List of extracted data from the object store based on search parameters
+    """
+
+    if filename:
+        search_df = read_input_file(filename)
+
+    if search_df is None:
+        raise ValueError("Either filename or extracted search dataframe must be supplied to retrieve data")
+
+    obs_type_name = "obs_type"
+    default_obs_type = "surface-insitu"
+    if obs_type_name not in search_df.columns:
+        logger.warning(
+            f"No 'obs_type' column has been supplied within the search parameters. Defaulting to: '{default_obs_type}'"
+        )
+        search_df[obs_type_name] = default_obs_type
+
+    get_functions, get_fn_arguments = define_get_functions()
+
+    data_object_all = []
+    for i, row in search_df.iterrows():
+        obs_type = row[obs_type_name]
+        get_fn = get_functions[obs_type]
+        additional_fn_arguments = get_fn_arguments[get_fn]
+
+        kwargs = row.to_dict()
+        kwargs.pop(obs_type_name)
+
+        if store:
+            kwargs["store"] = store
+
+        # TODO: Update to a more robust check (may be code to already do this?)
+        if "-" in kwargs["inlet"]:
+            start, end = kwargs["inlet"].split("-")
+            kwargs["inlet"] = slice(start, end)
+
+        # Remove any NaN entries
+        # TODO: Decide how this could work with negative lookup?
+        kwargs = {key: value for key, value in kwargs.items() if not pd.isnull(value)}
+
+        # Pass any additional arguments needed for the get/search function
+        kwargs = kwargs | additional_fn_arguments
+
+        # TODO: Decide details around catching errors for no files/multi files found.
+        data_retrieved = get_fn(**kwargs)
+
+        DataRetrieved = StoredData(data_retrieved, obs_type=obs_type)
```

Comment:
This should adhere to "lower snake case" format, so maybe `stored_data` or `_data_retrieved`.

Reply:
Ah yes you're right - because this a list not a class itself:

```suggestion
        stored_data = StoredData(data_retrieved, obs_type=obs_type)
```